### DISCUSSION TITLE
Struggling to implement markdoc-example tags

### DISCUSSION THREAD

**Discussion by NiallRees** [markdoc/markdoc#304] [Q&A]
Hi there, I'm trying to add a code block to a markdoc file. Inside the code block is some jinja, which is purely for demonstration purposes (I don't want the jinja to render). For example:

````
{% markdoc-example %}
```sql
{{ config(
    snowflake_warehouse=get_warehouse('large')
) }}

select
	...
from {{ ref('stg_orders') }}
```
{% /markdoc-example %}
````

I have configured the following in `tags.js` ([following this](https://github.com/markdoc/docs/blob/main/markdoc/tags/markdoc-example.markdoc.js)):
```
  'markdoc-example': {
    render: InlineCode,
    attributes: {},
    transform(node, config) {
      const attributes = node.transformAttributes(config)
      const { content, language } = node.children[0].attributes
      return new Tag(this.render, { ...attributes, language }, [content])
    },
  },
```

But when trying to build the docs, I'm getting the following error:
````
Error: Node 'markdoc-example' is missing closing
{% markdoc-example %}

Node 'fence' is missing closing
```sql
{% macro get_warehouse(size) %}

    {% set available_sizes = ['xsmall', 'small', 'medium', 'large', 'xlarge', '2xlarge'] %}
    {% if size not in available_sizes %}
        {{ exceptions.raise_compiler_error("Warehouse size not one of " ~ valid_warehouse_sizes) }}
    {% endif %}
    {% if target.name in ('production', 'prod') %}
                                {% do return('dbt_production_' ~ size) %}
    {% elif target.name in ('ci') %}
        {% do return('dbt_ci_' ~ size) %}
    {% else %}
        {% do return(None) %}
    {% endif %}
{% endmacro %}
```

Expected "(" or "=" but ")" found.
{% macro get_warehouse(size) %}

Expected "(" or "=" but " " found.
    {% set available_sizes = ['xsmall', 'small', 'medium', 'large', 'xlarge', '2xlarge'] %}

Expected "(" or "=" but " " found.
    {% if size not in available_sizes %}

Node 'endif' is missing closing
    {% endif %}

Expected "(" or "=" but "." found.
    {% if target.name in ('production', 'prod') %}

Expected ")", ",", "[", "{", boolean, identifier, null, number, string, variable, or whitespace but "'" found.
                                {% do return('dbt_production_' ~ size) %}

Expected "(" or "=" but "." found.
    {% elif target.name in ('ci') %}

Expected ")", ",", "[", "{", boolean, identifier, null, number, string, variable, or whitespace but "'" found.
        {% do return('dbt_ci_' ~ size) %}

Node 'else' is missing closing
    {% else %}

'else' tag should be self-closing
    {% else %}

Expected "(" or "=" but ")" found.
        {% do return(None) %}
````

I'm still very new to markdoc, so it's likely I've missed something obvious! Appreciate any help.

**Comment by NiallRees** [comment]
Now I'm trying to use the `{% process=false %}` flag per the docs, but every fence that I've added that to is now empty after rendering:
<img width="999" alt="image" src="https://user-images.githubusercontent.com/23722609/209239439-6c838644-5bf4-4af8-9a06-7e248a74e8ff.png">

**Reply by rpaul-stripe** [reply]
This is technically the correct answer. If you want the templates to pass through as plain text, using `process=false` on the fence prevents them from parsing. I'd like to try to troubleshoot why that wasn't working for you, but I'm having a little trouble reproducing your issue. When I put this content into a minimal example with the HTML renderer, it works as expected.

```javascript
import Markdoc from '@markdoc/markdoc';

const example = `
~~~sql {% process=false %}
{% macro get_warehouse(size) %}

    {% set available_sizes = ['xsmall', 'small', 'medium', 'large', 'xlarge', '2xlarge'] %}
    {% if size not in available_sizes %}
        {{ exceptions.raise_compiler_error("Warehouse size not one of " ~ valid_warehouse_sizes) }}
    {% endif %}
    {% if target.name in ('production', 'prod') %}
                                {% do return('dbt_production_' ~ size) %}
    {% elif target.name in ('ci') %}
        {% do return('dbt_ci_' ~ size) %}
    {% else %}
        {% do return(None) %}
    {% endif %}
{% endmacro %}
~~~
`;

const doc = Markdoc.parse(example);
const output = Markdoc.transform(doc, {});
const html = Markdoc.renderers.html(output);
console.log(html);
```

I get the following output, and no validation errors:

```
<article><pre data-language="sql">{% macro get_warehouse(size) %}

    {% set available_sizes = ['xsmall', 'small', 'medium', 'large', 'xlarge', '2xlarge'] %}
    {% if size not in available_sizes %}
        {{ exceptions.raise_compiler_error(&quot;Warehouse size not one of &quot; ~ valid_warehouse_sizes) }}
    {% endif %}
    {% if target.name in ('production', 'prod') %}
                                {% do return('dbt_production_' ~ size) %}
    {% elif target.name in ('ci') %}
        {% do return('dbt_ci_' ~ size) %}
    {% else %}
        {% do return(None) %}
    {% endif %}
{% endmacro %}
</pre></article>
```

Are you using a custom schema definition for the `fence` node?

**Comment by NiallRees** [comment]
Would love some help here @mfix-stripe - thanks a bunch

**Reply by mfix-stripe** [reply]
Sorry for the delay @NiallRees

First of all, you shouldn't need a `markdoc-example` tag just to render the first code block above. You can see from this example that just using a fence without it works: https://markdoc.dev/sandbox?mode=preview&c=YGBgc3FsCnt7IGNvbmZpZygKICAgIHNub3dmbGFrZV93YXJlaG91c2U9Z2V0X3dhcmVob3VzZSgnbGFyZ2UnKQopIH19CgpzZWxlY3QKCS4uLgpmcm9tIHt7IHJlZignc3RnX29yZGVycycpIH19CmBgYAo%253D (if you are curious for what is happening behind the scenes, this entire site is open source here: https://github.com/markdoc/docs)

However, seems like there might be an issue with the codeblock containing `{% macro get_warehouse(size) %}` — can you share the full contents of that Markdoc file so I can try and reproduce?

**Reply by ian-whitestone** [reply]
hey @mfix-stripe, this is offending codeblock:

```sql
{% macro get_warehouse(size) %}
    {% set available_sizes = ['xsmall', 'small', 'medium', 'large', 'xlarge', '2xlarge'] %}
    {% if size not in available_sizes %}
        {{ exceptions.raise_compiler_error("Warehouse size not one of " ~ valid_warehouse_sizes) }}
    {% endif %}
    {% if target.name in ('production', 'prod') %}
				{% do return('dbt_production_' ~ size) %}
    {% elif target.name in ('ci') %}
        {% do return('dbt_ci_' ~ size) %}
    {% else %}
        {% do return(None) %}
    {% endif %}
{% endmacro %}
``` 


so it looks like this:



```
{% markdoc-example %}
``sql
{% macro get_warehouse(size) %}
    {% set available_sizes = ['xsmall', 'small', 'medium', 'large', 'xlarge', '2xlarge'] %}
    {% if size not in available_sizes %}
        {{ exceptions.raise_compiler_error("Warehouse size not one of " ~ valid_warehouse_sizes) }}
    {% endif %}
    {% if target.name in ('production', 'prod') %}
				{% do return('dbt_production_' ~ size) %}
    {% elif target.name in ('ci') %}
        {% do return('dbt_ci_' ~ size) %}
    {% else %}
        {% do return(None) %}
    {% endif %}
{% endmacro %}
`` (actually three ticks instead of two)
{% /markdoc-example %}
```

**Reply by ian-whitestone** [reply]
> First of all, you shouldn't need a markdoc-example tag just to render the first code block above. You can see from this example that just using a fence without it works

I believe what happened is:

1. The default fence didn't work when the codeblock contained jinja
2. @NiallRees saw https://github.com/markdoc/markdoc/discussions/261, then tried implementing that as a fix

**Reply by mfix-stripe** [reply]
I have my `markdoc-tag` tag defined as:
```js
{
  render: Code,
  attributes: {},
  transform(node, config) {
    const attributes = node.transformAttributes(config);
    const { content, language } = node.children[0].attributes;

    return new Tag(this.render, { ...attributes, language }, [content]);
  }
}
```
as you can see here: https://github.com/markdoc/docs/blob/main/markdoc/tags/markdoc-example.markdoc.js

which works for me: https://markdoc.dev/sandbox?mode=preview&c=eyUgbWFya2RvYy1leGFtcGxlICV9CmBgYHNxbAp7JSBtYWNybyBnZXRfd2FyZWhvdXNlKHNpemUpICV9CiAgICB7JSBzZXQgYXZhaWxhYmxlX3NpemVzID0gWyd4c21hbGwnLCAnc21hbGwnLCAnbWVkaXVtJywgJ2xhcmdlJywgJ3hsYXJnZScsICcyeGxhcmdlJ10gJX0KICAgIHslIGlmIHNpemUgbm90IGluIGF2YWlsYWJsZV9zaXplcyAlfQogICAgICAgIHt7IGV4Y2VwdGlvbnMucmFpc2VfY29tcGlsZXJfZXJyb3IoIldhcmVob3VzZSBzaXplIG5vdCBvbmUgb2YgIiB%252BIHZhbGlkX3dhcmVob3VzZV9zaXplcykgfX0KICAgIHslIGVuZGlmICV9CiAgICB7JSBpZiB0YXJnZXQubmFtZSBpbiAoJ3Byb2R1Y3Rpb24nLCAncHJvZCcpICV9CgkJCQl7JSBkbyByZXR1cm4oJ2RidF9wcm9kdWN0aW9uXycgfiBzaXplKSAlfQogICAgeyUgZWxpZiB0YXJnZXQubmFtZSBpbiAoJ2NpJykgJX0KICAgICAgICB7JSBkbyByZXR1cm4oJ2RidF9jaV8nIH4gc2l6ZSkgJX0KICAgIHslIGVsc2UgJX0KICAgICAgICB7JSBkbyByZXR1cm4oTm9uZSkgJX0KICAgIHslIGVuZGlmICV9CnslIGVuZG1hY3JvICV9CmBgYAp7JSAvbWFya2RvYy1leGFtcGxlICV9

I do still see a validation error, but the content renders correctly, so these errors can probably be ignored.

**Reply by mfix-stripe** [reply]
@rpaul-stripe do we do anything special to ignore these parsing errors when inside a `markdoc-example` tag?

**Reply by NiallRees** [reply]
Thanks @mfix-stripe. I'm unfortunately still getting an error using:

````
{% markdoc-example %}

```sql
{% macro get_warehouse(size) %}
    {% set available_sizes = ['xsmall', 'small', 'medium', 'large', 'xlarge', '2xlarge'] %}
    {% if size not in available_sizes %}
        {{ exceptions.raise_compiler_error("Warehouse size not one of " ~ valid_warehouse_sizes) }}
    {% endif %}
    {% if target.name in ('production', 'prod') %}
				{% do return('dbt_production_' ~ size) %}
    {% elif target.name in ('ci') %}
        {% do return('dbt_ci_' ~ size) %}
    {% else %}
        {% do return(None) %}
    {% endif %}
{% endmacro %}
```

{% /markdoc-example %}
````

My `markdoc-example` is defined as:

```
  'markdoc-example': {
    render: InlineCode,
    attributes: {},
    transform(node, config) {
      const attributes = node.transformAttributes(config)
      const { content, language } = node.children[0].attributes
      return new Tag(this.render, { ...attributes, language }, [content])
    },
  },
```

The error:
````
./src/pages/posts/configuring-snowflake-warehouse-sizes-in-dbt.md
Error: Node 'markdoc-example' is missing closing
{% markdoc-example %}

Node 'fence' is missing closing
```sql
{% macro get_warehouse(size) %}
    {% set available_sizes = ['xsmall', 'small', 'medium', 'large', 'xlarge', '2xlarge'] %}
    {% if size not in available_sizes %}
        {{ exceptions.raise_compiler_error("Warehouse size not one of " ~ valid_warehouse_sizes) }}
    {% endif %}
    {% if target.name in ('production', 'prod') %}
				{% do return('dbt_production_' ~ size) %}
    {% elif target.name in ('ci') %}
        {% do return('dbt_ci_' ~ size) %}
    {% else %}
        {% do return(None) %}
    {% endif %}
{% endmacro %}
```

Expected "(" or "=" but ")" found.
{% macro get_warehouse(size) %}

Expected "(" or "=" but " " found.
    {% set available_sizes = ['xsmall', 'small', 'medium', 'large', 'xlarge', '2xlarge'] %}

Expected "(" or "=" but " " found.
    {% if size not in available_sizes %}

Node 'endif' is missing closing
    {% endif %}

Expected "(" or "=" but "." found.
    {% if target.name in ('production', 'prod') %}

Expected ")", ",", "[", "{", boolean, identifier, null, number, string, variable, or whitespace but "'" found.
				{% do return('dbt_production_' ~ size) %}

Expected "(" or "=" but "." found.
    {% elif target.name in ('ci') %}

Expected ")", ",", "[", "{", boolean, identifier, null, number, string, variable, or whitespace but "'" found.
        {% do return('dbt_ci_' ~ size) %}

Node 'else' is missing closing
    {% else %}

'else' tag should be self-closing
    {% else %}

Expected "(" or "=" but ")" found.
        {% do return(None) %}
````

**Reply by mfix-stripe** [reply]
@NiallRees — yeah I mentioned that above. I am also still seeing validation errors, but the content will still render appropriately ([see link above](https://markdoc.dev/sandbox?mode=preview&c=eyUgbWFya2RvYy1leGFtcGxlICV9CmBgYHNxbAp7JSBtYWNybyBnZXRfd2FyZWhvdXNlKHNpemUpICV9CiAgICB7JSBzZXQgYXZhaWxhYmxlX3NpemVzID0gWyd4c21hbGwnLCAnc21hbGwnLCAnbWVkaXVtJywgJ2xhcmdlJywgJ3hsYXJnZScsICcyeGxhcmdlJ10gJX0KICAgIHslIGlmIHNpemUgbm90IGluIGF2YWlsYWJsZV9zaXplcyAlfQogICAgICAgIHt7IGV4Y2VwdGlvbnMucmFpc2VfY29tcGlsZXJfZXJyb3IoIldhcmVob3VzZSBzaXplIG5vdCBvbmUgb2YgIiB%252BIHZhbGlkX3dhcmVob3VzZV9zaXplcykgfX0KICAgIHslIGVuZGlmICV9CiAgICB7JSBpZiB0YXJnZXQubmFtZSBpbiAoJ3Byb2R1Y3Rpb24nLCAncHJvZCcpICV9CgkJCQl7JSBkbyByZXR1cm4oJ2RidF9wcm9kdWN0aW9uXycgfiBzaXplKSAlfQogICAgeyUgZWxpZiB0YXJnZXQubmFtZSBpbiAoJ2NpJykgJX0KICAgICAgICB7JSBkbyByZXR1cm4oJ2RidF9jaV8nIH4gc2l6ZSkgJX0KICAgIHslIGVsc2UgJX0KICAgICAgICB7JSBkbyByZXR1cm4oTm9uZSkgJX0KICAgIHslIGVuZGlmICV9CnslIGVuZG1hY3JvICV9CmBgYAp7JSAvbWFya2RvYy1leGFtcGxlICV9))

Checking w/ @rpaul-stripe if we do anything specific to ignore these validation errors.

**Reply by NiallRees** [reply]
@mfix-stripe thanks, unfortunately I'm seeing this on the page though:

> This error occurred during the build process and can only be dismissed by fixing the error.

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/23722609/212987940-0ebe1f05-4e50-4724-b067-ba2fb26522f9.png">

**Reply by mfix-stripe** [reply]
@NiallRees — which version of `@markdoc/markdoc` and `@markdoc/next.js` do you have installed?

**Reply by NiallRees** [reply]
Ok hallelujah it's now working, thanks for the help @mfix-stripe. I updated to the latest version which removed that error.

**Reply by mfix-stripe** [reply]
@NiallRees no problem! 

We still should figure out a way to remove those validation errors from `Markdoc.validate` though (Cc @rpaul-stripe)

**Reply by NiallRees** [reply]
Another issue has cropped up unfortunately @mfix-stripe - anything after the example tags isn't rendered. [See example](https://markdoc.dev/sandbox?mode=preview&c=SGkgdGhlcmUKCnslIG1hcmtkb2MtZXhhbXBsZSAlfQpgYGBzcWwKeyUgbWFjcm8gZ2V0X3dhcmVob3VzZShzaXplKSAlfQogICAgeyUgc2V0IGF2YWlsYWJsZV9zaXplcyA9IFsneHNtYWxsJywgJ3NtYWxsJywgJ21lZGl1bScsICdsYXJnZScsICd4bGFyZ2UnLCAnMnhsYXJnZSddICV9CiAgICB7JSBpZiBzaXplIG5vdCBpbiBhdmFpbGFibGVfc2l6ZXMgJX0KICAgICAgICB7eyBleGNlcHRpb25zLnJhaXNlX2NvbXBpbGVyX2Vycm9yKCJXYXJlaG91c2Ugc2l6ZSBub3Qgb25lIG9mICIgfiB2YWxpZF93YXJlaG91c2Vfc2l6ZXMpIH19CiAgICB7JSBlbmRpZiAlfQogICAgeyUgaWYgdGFyZ2V0Lm5hbWUgaW4gKCdwcm9kdWN0aW9uJywgJ3Byb2QnKSAlfQoJCQkJeyUgZG8gcmV0dXJuKCdkYnRfcHJvZHVjdGlvbl8nIH4gc2l6ZSkgJX0KICAgIHslIGVsaWYgdGFyZ2V0Lm5hbWUgaW4gKCdjaScpICV9CiAgICAgICAgeyUgZG8gcmV0dXJuKCdkYnRfY2lfJyB%252BIHNpemUpICV9CiAgICB7JSBlbHNlICV9CiAgICAgICAgeyUgZG8gcmV0dXJuKE5vbmUpICV9CiAgICB7JSBlbmRpZiAlfQp7JSBlbmRtYWNybyAlfQpgYGAKeyUgL21hcmtkb2MtZXhhbXBsZSAlfQoKIyMgSW1wb3J0YW50IHNlY3Rpb24KT3RoZXIgc3R1ZmYgSSB3YW50IHRvIHRhbGsgYWJvdXQ%253D). I'm guessing this is related to the missing closing tags error?

**Reply by NiallRees** [reply]
Ah ok - [now using process=false fixes it](https://markdoc.dev/sandbox?mode=preview&c=SGkgdGhlcmUKCnslIG1hcmtkb2MtZXhhbXBsZSAlfQpgYGBzcWwgeyUgcHJvY2Vzcz1mYWxzZSAlfQp7JSBtYWNybyBnZXRfd2FyZWhvdXNlKHNpemUpICV9CiAgICB7JSBzZXQgYXZhaWxhYmxlX3NpemVzID0gWyd4c21hbGwnLCAnc21hbGwnLCAnbWVkaXVtJywgJ2xhcmdlJywgJ3hsYXJnZScsICcyeGxhcmdlJ10gJX0KICAgIHslIGlmIHNpemUgbm90IGluIGF2YWlsYWJsZV9zaXplcyAlfQogICAgICAgIHt7IGV4Y2VwdGlvbnMucmFpc2VfY29tcGlsZXJfZXJyb3IoIldhcmVob3VzZSBzaXplIG5vdCBvbmUgb2YgIiB%252BIHZhbGlkX3dhcmVob3VzZV9zaXplcykgfX0KICAgIHslIGVuZGlmICV9CiAgICB7JSBpZiB0YXJnZXQubmFtZSBpbiAoJ3Byb2R1Y3Rpb24nLCAncHJvZCcpICV9CgkJCQl7JSBkbyByZXR1cm4oJ2RidF9wcm9kdWN0aW9uXycgfiBzaXplKSAlfQogICAgeyUgZWxpZiB0YXJnZXQubmFtZSBpbiAoJ2NpJykgJX0KICAgICAgICB7JSBkbyByZXR1cm4oJ2RidF9jaV8nIH4gc2l6ZSkgJX0KICAgIHslIGVsc2UgJX0KICAgICAgICB7JSBkbyByZXR1cm4oTm9uZSkgJX0KICAgIHslIGVuZGlmICV9CnslIGVuZG1hY3JvICV9CmBgYAp7JSAvbWFya2RvYy1leGFtcGxlICV9CgojIyBJbXBvcnRhbnQgc2VjdGlvbgpPdGhlciBzdHVmZiBJIHdhbnQgdG8gdGFsayBhYm91dA%253D%253D)
